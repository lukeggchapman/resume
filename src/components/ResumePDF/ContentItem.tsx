import React from 'react'
import { View, Text, StyleSheet, Image } from '@react-pdf/renderer'
import { differenceInMonths, addDays, format } from 'date-fns'

import hr from '../../images/hr.png'

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    fontSize: 10,
    padding: '0 5',
  },
  leftColumn: {
    flex: 1,
    alignItems: 'center',
    justifyItems: 'center',
    paddingRight: 5,
  },
  rightColumn: {
    flex: 7,
  },
  title: {
    fontFamily: 'Oswald',
    fontSize: 14,
  },
  subTitle: {
    fontFamily: 'PT Sans',
    fontSize: 10,
    color: '#555',
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  contentContainer: {
    flexDirection: 'column',
    width: '100%',
  },
  content: {
    marginTop: 5,
  },
  logo: {
    height: 35,
  },
  hr: {
    width: 400,
    alignSelf: 'center',
    margin: '10 0',
  },
  dates: {
    flexDirection: 'row',
  },
})

const getDifference = (startDate: string, endDate: string) => {
  const endDateAdd1 = addDays(endDate, 1) // Most contracts will end on the last day of the month
  const total = differenceInMonths(endDateAdd1, startDate)
  const years = Math.floor(total / 12)
  const months = total % 12

  return [
    (years && `${years} yr${years > 1 ? 's' : ''}`) || undefined,
    (months && `${months} mo${months > 1 ? 's' : ''}`) || undefined,
  ]
    .filter(string => !!string)
    .join(' ')
}

interface ContentItemProps {
  title: string
  subTitle: string
  logo?: string
  startDate: string
  endDate: string
  children?: React.ReactNode
}

const ContentItem = ({
  title,
  subTitle,
  logo,
  startDate,
  endDate,
  children,
}: ContentItemProps) => (
  <View style={styles.container}>
    <View style={styles.leftColumn}>
      {logo && <Image style={styles.logo} src={logo} />}
    </View>
    <View
      style={[
        styles.rightColumn,
        { alignSelf: children ? 'normal' : 'center' },
      ]}
    >
      <Text style={styles.title}>{title}</Text>
      <View style={styles.subTitle}>
        <Text>{subTitle}</Text>
        <View>
          <Text>
            {`${format(startDate, 'MMM, YYYY')} \u2013 ${format(
              endDate,
              'MMM, YYYY'
            )}`}
            {` \u00B7 `}
            {getDifference(startDate, endDate)}
          </Text>
        </View>
      </View>
      <View style={styles.contentContainer}>
        {children && <View style={styles.content}>{children}</View>}
        <Image style={styles.hr} src={hr} />
      </View>
    </View>
  </View>
)

export default ContentItem
