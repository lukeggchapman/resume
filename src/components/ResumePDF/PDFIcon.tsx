import React from 'react'
import { Text, StyleSheet } from '@react-pdf/renderer'

import './pdfFonts/fontawesome'
import fa from 'utils/fontawesome'

const styles = StyleSheet.create({
  icon: {
    fontFamily: 'Font Awesome 5 Free',
  },
  fab: {
    fontWeight: 400,
  },
  far: {
    fontWeight: 400,
  },
  fas: {
    fontWeight: 900,
  },
})

type IconStyleProps =
  | {
      fab: true
      far?: false
      fas?: false
    }
  | {
      fab?: false
      far: true
      fas?: false
    }
  | {
      fab?: false
      far?: false
      fas: true
    }

interface PDFIconProps {
  icon: keyof typeof fa
}

export default ({ icon, fab, far, fas }: PDFIconProps & IconStyleProps) => {
  const faStyle =
    (fab && styles.fab) ||
    (far && styles.far) ||
    (fas && styles.fas) ||
    styles.far

  return <Text style={[styles.icon, faStyle]}>{fa[icon]}</Text>
}
