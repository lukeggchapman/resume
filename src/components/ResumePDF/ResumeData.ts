import { Root } from '../../utils/htmlToPDF/hast.d'

export interface ExperienceData {
  edges: {
    node: {
      id: string
      htmlAst: Root
      frontmatter: {
        role: string
        company: string
        startDate: string
        endDate: string
        startDateFormatted: string
        endDateFormatted: string
        logo?: {
          publicURL: string
        }
      }
    }
  }[]
}

export default interface ResumeData {
  experience: ExperienceData
}
