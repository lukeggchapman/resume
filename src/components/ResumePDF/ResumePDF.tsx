import React from 'react'
import { css } from '@emotion/core'
import { PDFViewer } from '@react-pdf/renderer'

import Document from './components/Document'
import { ResumeData } from './resumeQuery'

export interface PureResumePDFProps {
  pageContext: {
    resumeData: ResumeData
  }
}

// Used in development only, client side renders PDF
const ResumePDF = ({ pageContext: { resumeData } }: PureResumePDFProps) => {
  return (
    <PDFViewer
      css={css`
        width: 100%;
        height: 100%;
        background-color: white;
        overflow-x: hidden;
        overflow-y: hidden;
        margin-bottom: 0;
        border: none;
      `}
    >
      <Document data={resumeData} />
    </PDFViewer>
  )
}

export default ResumePDF
