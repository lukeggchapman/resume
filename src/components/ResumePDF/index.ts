export { default } from './ResumePDF'
export * from './ResumePDF'
