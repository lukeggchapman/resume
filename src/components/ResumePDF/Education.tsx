import React from 'react'

import ContentItem from './ContentItem'
import macquarieUniLogo from '../../images/macquarie-uni.png'

const Experience = () => (
  <ContentItem
    title="Macquarie University"
    subTitle="Bachelor of Information Technology"
    logo={macquarieUniLogo}
    startDate="2004-01-01"
    endDate="2007-12-31"
  />
)

export default Experience
