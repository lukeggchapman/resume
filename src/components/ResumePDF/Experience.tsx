import React from 'react'
import { View, Text, StyleSheet } from '@react-pdf/renderer'

import { ExperienceData } from './ResumeData'
import htmlToPDF from '../../utils/htmlToPDF'

interface ExperienceProps {
  data: ExperienceData
}

const styles = StyleSheet.create({
  content: {
    fontFamily: 'PT Sans',
    fontSize: 10,
  },
})

const sortGroupsByEndDate = (group: ExperienceData['group']) => {
  return group.sort((a, b) =>
    a.edges[0].node.frontmatter.endDate > b.edges[0].node.frontmatter.endDate
      ? 1
      : 0
  )
}

const Experience = ({ data }: ExperienceProps) => {
  const groupSorted = sortGroupsByEndDate(data.group)

  return (
    <>
      {groupSorted.map(group => {
        const { company } = group.edges[0].node.frontmatter

        return (
          <View key={company}>
            <Text>{company}</Text>
            {group.edges.map(edge => {
              const experience = edge.node
              const {
                endDateFormatted,
                startDate,
                endDate,
              } = experience.frontmatter

              return (
                <View style={styles.content} key={`${company}_${endDate}`}>
                  <Text>{`${startDate} - ${endDateFormatted}`}</Text>
                  <View>{htmlToPDF(experience.htmlAst)}</View>
                </View>
              )
            })}
          </View>
        )
      })}
    </>
  )
}

export default Experience
