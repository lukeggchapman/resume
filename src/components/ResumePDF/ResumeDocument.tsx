import React from 'react'
import {
  Page,
  Text,
  Link as PDFLink,
  View,
  Document,
  StyleSheet,
  Font,
} from '@react-pdf/renderer'
import PDFIcon from './PDFIcon'

import ResumeData from './ResumeData'
import Experience from './Experience'
import Education from './Education'
import Section from './Section'

interface ResumePDFProps {
  data: ResumeData
}

Font.register({
  family: 'Oswald',
  src: 'https://fonts.gstatic.com/s/oswald/v13/Y_TKV6o8WovbUd3m_X9aAA.ttf',
})

Font.register({
  family: 'PT Sans',
  src: 'http://fonts.gstatic.com/s/ptsans/v8/FUDHvzEKSJww3kCxuiAo2A.ttf',
})

const stylesObj = {
  page: {
    flexDirection: 'column' as 'column',
    backgroundColor: 'white',
    fontFamily: 'PT Sans',
    fontSize: 12,
    color: '#333',
  },
  header: {
    flexDirection: 'row' as 'row',
    padding: '2% 3%',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  heading: {
    fontFamily: 'Oswald',
    fontWeight: 700,
    fontSize: 24,
  },
  info: {
    alignItems: 'right',
  },
  body: {
    flexDirection: 'row',
    flexGrow: 1,
    padding: 10,
  },
  sidePanel: {
    flex: 1,
  },
  content: {
    flex: 3,
  },
  footer: {},
}
const styles = StyleSheet.create(stylesObj)

const ResumeDocument = ({ data }: ResumePDFProps) => {
  return (
    <Document>
      <Page size="A4" style={styles.page}>
        <View style={styles.header}>
          <Text style={styles.heading}>Luke Chapman</Text>
          <View style={styles.info}>
            <Text>
              <PDFIcon icon="envelope" fas />
              {` `}
              <PDFLink src="mailto:lukeggchapman@gmail.com">
                LukeGGChapman@gmail.com
              </PDFLink>
            </Text>
            <Text>
              <PDFIcon icon="phone" fas />
              {` `}
              <PDFLink src="tel:+61401516361">+61 401 516 361</PDFLink>
            </Text>
          </View>
        </View>
        <View style={styles.body}>
          <View style={styles.sidePanel}>
            <Text>Side Panel</Text>
          </View>
          <View style={styles.content}>
            <Section title="Education">
              <Education />
            </Section>
            <Section title="Experience">
              <Experience data={data.experience} />
            </Section>
          </View>
        </View>
        <View style={styles.footer}>
          <Text>Footer</Text>
        </View>
      </Page>
    </Document>
  )
}

export default ResumeDocument
