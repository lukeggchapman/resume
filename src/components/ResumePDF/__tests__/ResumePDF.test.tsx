import React from 'react'
import { shallow } from 'enzyme'

import ResumePDF, { PureResumePDFProps } from '../ResumePDF'

jest.mock('../components/Document', () => 'Document')

const baseProps: PureResumePDFProps = {
  pageContext: {
    resumeData: 'dataTest' as any,
  },
}

describe('ResumePDF', () => {
  it('renders correctly', () => {
    expect(shallow(<ResumePDF {...baseProps} />)).toMatchSnapshot()
  })
})
