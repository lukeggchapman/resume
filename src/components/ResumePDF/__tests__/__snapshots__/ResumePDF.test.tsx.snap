// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResumePDF renders correctly 1`] = `
<PDFViewer
  css={
    Object {
      "map": undefined,
      "name": "dts0fx",
      "next": undefined,
      "styles": "
        width: 100%;
        height: 100%;
        background-color: white;
        overflow-x: hidden;
        overflow-y: hidden;
        margin-bottom: 0;
        border: none;
      ",
    }
  }
>
  <Document
    data="dataTest"
  />
</PDFViewer>
`;
